s3sync.go

```diff
@@ -354,6 +364,74 @@ func (m *Manager) deleteRemote(file *fileInfo, destPath *s3Path) error {
 	return err
 }
 
+func (m *Manager) softDeleteRemote(file *fileInfo, destPath *s3Path) error {
+	destFile := *destPath
+	if strings.HasSuffix(destPath.bucketPrefix, "/") || destPath.bucketPrefix == "" || !file.singleFile {
+		// If source is a single file and destination is not a directory, use destination URL as is.
+		destFile.bucketPrefix = filepath.Join(destPath.bucketPrefix, file.name)
+	}
+
+	head, err := m.s3.HeadObject(&s3.HeadObjectInput{
+		Bucket: aws.String(destFile.bucket),
+		Key:    aws.String(destFile.bucketPrefix),
+	})
+	if err != nil {
+		return err
+	}
+
+	if m.dryrun {
+		return nil
+	}
+
+	if head.Expires != nil {
```

Comment:
Do we need to separately handle this? In other words is there a way to just make s3 handle the deletion

Reply:
AWS S3 does have this concept of "lifecycle rules" which can handle expiration, but it's extremely blunt and can basically only handle deleting objects (with key prefix targeting) by making a decision on the objects creation date and the configured lifecycle period. To me that seems a bit too rough outside of the scope of e.g. log files or such 😰 .